napari/util/keybindings.py

```diff
@@ -275,98 +284,135 @@ class KeybindingDescriptor:
-        try:
-            self[key]
-        except KeyError:
-            return False
-        else:
-            return True
 
-    def __delitem__(self, key):
-        self[key] = None
+class KeymapHandler:
+    """Mix-in to add key handling functionality.
 
-    def copy(self):
-        return InheritedKeymap(self._parent, self.data.copy())
+    Attributes
+    ----------
+    keymap_providers : list of KeymapProvider
+        Classes that provide the keymaps for this class to handle.
+    """
 
-    def pop(self, key, default=UNDEFINED):
-        try:
-            default = self[key]
-        except KeyError:
-            if default is UNDEFINED:
-                raise
-        else:
-            del self[key]
+    def __init__(self):
+        super().__init__()
+        self._key_release_generators = {}
+        self.keymap_providers = []
 
-        return default
+    @property
+    def keymap_chain(self):
+        """collections.ChainMap: Chain of keymaps from keymap providers."""
+        maps = []
 
+        for parent in self.keymap_providers:
+            maps.append(parent.keymap)
+            maps.append(parent.class_keymap)
 
-class KeymapMixin:
-    """Mix-in to add keymap functionality. Must still define ``class_keymap``
-    in every subclass.
-    """
+        return ChainMap(*maps)
 
-    def __init__(self):
-        self.keymap = {}
+    @property
+    def active_keymap(self):
+        """dict: Active keymap, created by resolving the keymap chain."""
+        keymaps = self.keymap_chain.maps
 
-    def __init_subclass__(cls, **kwargs):
-        super().__init_subclass__(**kwargs)
+        for i, keymap in enumerate(keymaps):
+            if Ellipsis in keymap:  # catch-all
+                break
+        keymaps = keymaps[: i + 1]  # trim maps after a catch-all
 
-        if 'class_keymap' not in cls.__dict__:
-            # if in __dict__, was defined in class and not inherited
-            cls.class_keymap = {}
+        active_keymap = {}
 
-    @property
-    def keymap(self):
-        """InheritedKeymap : Keymap used for shortcuts.
-        Inherits from ``class_keymap``.
+        for keymap in reversed(keymaps):
+            for key, func in keymap.items():
+                active_keymap[key] = func
+                if func is Ellipsis:  # blocker
+                    del active_keymap[key]
 
-        Do not directly set key bindings; use ``bind_key`` instead.
-        """
-        return self._keymap.copy()
+        return active_keymap
 
-    @keymap.setter
-    def keymap(self, keymap):
-        self._keymap = InheritedKeymap(lambda: self.class_keymap, keymap)
+    def press_key(self, key_combo):
+        """Simulate a key press to activate a keybinding.
 
-    bind_key = KeybindingDescriptor(bind_key)
+        Parameters
+        ----------
+        key_combo : str
+            Key combination.
+        """
+        key_combo = normalize_key_combo(key_combo)
+
+        for parent in self.keymap_providers:
```

Comment:
> Can the parent be somehow bound to the function when the shortcut is created?

Ohhhh I like this! I think I'll put this in the `keymap_chain` property :)

> If we are concerned with creating the whole thing each time (which we are), this could be solved by caching + cache invalidation...

Hmmm I'm not quite sure what the best way to do this would be. How about I just do the above and we can let someone more experienced work on the optimization?

Reply:
Haha I'm sure you are experienced enough, but yes, "get it running, get it right, get it fast" is a great mantra. =)

Reply:
but, essentially, the idea would be `self._cached_active_keymap` would be created, and anything that touches the keymap_chain property resets that to `None`. Then the `active_keymap` property first checks to see whether the cache is None, if it isn't, returns it, if it is, it does its magic.